### ISSUE TITLE
Why not use ArrayBuffer?

### ISSUE THREAD

**Issue by larsgk** [WICG/serial#41]
As someone who has been quite happy with how the chrome.serial API worked up until deprecation, I am wondering why this spec uses streams for communication (which seem to be in flux/not final?) while ArrayBuffers - used by chrome.serial - are easy to deal with and widely understood/supported.  Maybe I am missing something?

ref:

https://developer.chrome.com/apps/serial#method-send
https://developer.chrome.com/apps/serial#event-onReceive

**Comment by marcoscaceres** [comment]
Historically, when we started, serial sitting atop of streams made sense. FWIW, streams are in Chrome, but we've had to delay implementation in Firefox for this year (we are focusing of multiprocess Firefox and other stability goodies). 

People with experience in this area, it would be great to hear which primitive would be best to leverage (or if we should leverage Chrome's existing API wholesale).

**Comment by larsgk** [comment]
IMO, the chrome.serial API is very easy to understand and works well with the PnP (VID/PID/"FriendlyName") discovery mechanism they got in place eventually.

The only thing I would change for sure is the global listeners:

    chrome.serial.onReceive.addListener(onReceiveCallback);
    chrome.serial.onReceiveError.addListener(onSerialError);

Filtering on which connection sent what inside the listener seems much less clean than having a specific callback for *this* connection (e.g. returned in a promise on connect to a port)

**Comment by reconbot** [comment]
On the original topic of `ArrayBuffer` vs `Uint8Array` I don't yet have any informed opinions (even if my gut say `Uint8Array` because they look more friendly to developers most of the time). As long as they're not strings I'll be happy. 

On the side topic of copying chrome serialport that @marcoscaceres brings up;

It should be no surprise that I am partial to the node serialport api as I'm the current maintainer of that project. However I think chrome does it great overall. I love their well defined connection and configuration objects and control functions but I dislike the the DOM like aspects of receiving data and events and that weird global emitter.

I haven't done a proper review yet (I plan to) but I think we're not quite yet as well defined as node serialport or chrome serialport in terms of construction, configuration and control options and functions. 

As for data flow, I I love the Node.js style duplex stream. In particular reading and writing to the serial port object itself is a beautiful thing. This duplex stream isn't in the streams spec as far as I can tell and I don't know why.

The 5.x release node serialport will support the Node.js Streams 3 API. Which in addition to data events will give us the ability to request a specific number of bytes from the port. So you can read in a `pull` or `push` fashion. This is supported by the proposed streams api and is not the chrome serialport `onReceive` callback.

I think the internals of node streams are hard to define and test, it's not a panacea. Additionally Interoperability with other browser stream like apis seems like a great goal.

**Comment by marcoscaceres** [comment]
Regarding "duplex stream isn't in the streams spec as far as I can tell and I don't know why"? cc @domenic

**Comment by larsgk** [comment]
The node serialport project is great for what it's intended - but I must say, I prefer the listing and filtering method currently proposed in WebUSB - that would also prompt for user acceptance to get a promise successfully delivering the port (very similar to the approach proposed in the current version of this spec):

```
serial.requestPort = function() {
    const filters = [
      { 'vendorId': 0x2341, 'productId': 0x8036 },
      { 'vendorId': 0x2341, 'productId': 0x8037 },
    ];
    return navigator.usb.requestDevice({ 'filters': filters }).then(
      device => new serial.Port(device)
    );
  }
```

In any case, we need a VERY simple AND secure way of:

1. search and find based on VID/PID (or BT MAC - it may be impossible to support plain RS232 in a secure way - what were the thoughts on that?)
2. connect to and handle data flow to/from one or multiple ports in parallel with least misunderstanding for users and developers of the web apps

**Comment by fbender** [comment]
Please stay on topic. See e.g. #42 for VID/PID/MAC-related stuff. If this or other existing issues do not cover your comments, please take the time to create a new issue. Then you can link to the issue in a comment that still (mostly) stays on topic.

**Comment by larsgk** [comment]
@fbender you are right - I will keep it in mind

**Comment by fbender** [comment]
The Fetch API is gaining stream support while it already shipped with other capabilities incl. ArrayBuffer support. Would it make sense to go a similar route for WebSerial? I.e. spec out an ArrayBuffer + events API *and* a stream API? Are there use cases that benefit from either method more than the other? How much can we share between the two methods? 

Can't we tunnel ArrayBuffer types over streams? E.g. we could spec this and later extend/change to a lower level method using streams, which allows to "shim" the ArrayBuffer method on top of it?

**Comment by domenic** [comment]
> Regarding "duplex stream isn't in the streams spec as far as I can tell and I don't know why"?

Just use a { readable, writable } pair. Example: https://streams.spec.whatwg.org/#example-both

**Comment by reillyeon** [comment]
The current draft specification uses a `{ readable, writable }` pair. Chunks produced by the `ReadableStream` are always a `Uint8Array` and any `BufferSource` (`ArrayBuffer` or `ArrayBufferView`) is accepted as a chunk by the `WritableStream`.
